src/Microsoft.AspNetCore.OData/Batch/ODataBatchReaderExtensions.cs

```diff
@@ -239,17 +238,9 @@ private static HttpContext CreateHttpContext(HttpContext originalContext)
 
             features[typeof(IHttpResponseFeature)] = new HttpResponseFeature();
 
-            // Create a context from the factory or use the default context.
-            HttpContext context = null;
-            IHttpContextFactory httpContextFactory = originalContext.RequestServices.GetRequiredService<IHttpContextFactory>();
-            if (httpContextFactory != null)
-            {
-                context = httpContextFactory.Create(features);
-            }
-            else
-            {
-                context = new DefaultHttpContext(features);
-            }
+            // Create a context.
+            // IHttpContextFactory should not be used, because it resets IHttpContextAccessor.HttpContext;
+            HttpContext context = new DefaultHttpContext(features);
```

Comment:
>HttpContext context = new DefaultHttpContext(features); [](start = 12, length = 55)

What are the ramifications of not using the context factory?  Doesn't that mean that, if someone adds their own context factory, we don't use it for batch?  Couldn't that be a problem if the developer is expecting their custom httpcontext to be used?  Do we instead need a way to create the context from the factory without resetting the IHttpContextAccessor.HttpContext?

Reply:
I do not see any other reason to use the factory for creating `HttpContext` from a set of sub-requests, but dependency injection. And it is unclear why a global factory (that creates context for an actual http request) should be used to instance batch sub-contexts.

If the factory is used we should make a `closure` of original context after processing batch changesets like this one:
```csharp
        /// <inheritdoc />
        public override async Task ProcessBatchAsync(HttpContext context, RequestDelegate nextHandler)
        {
            // Retrieve current httpcontext.
            var httpContextAccessor = context.RequestServices.GetService<IHttpContextAccessor>();
            var beforeContext = httpContextAccessor?.HttpContext;

            await base.ProcessBatchAsync(context, nextHandler);

            if (httpContextAccessor != null)
            {
                httpContextAccessor.HttpContext = beforeContext;
            }
        }
```
But still a closure is more like a workaround than THE solution.

Keyword `new` is kinda bad in common, especially in the name of DI, instead we can introduce an interface `IBatchHttpContextFactory` to create a context of the changeset.

Reply:
Hi @marabooy
Yeah, we should give more control of the context creation. But as I've pointed out `IHttpContextFactory` is not a way to do so. I suggest a new abstraction `IBatchHttpContextFactory` for this matter.